master/internal/project/postgres_project.go

```diff
@@ -50,3 +136,285 @@ func ProjectIDByName(ctx context.Context, workspaceID int, projectName string) (
 	}
 	return &pID, nil
 }
+
+// GetProjectByKey returns a project using its key to identify it.
+func GetProjectByKey(ctx context.Context, key string) (*model.Project, error) {
+	project := &model.Project{}
+	err := db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		var projectID int
+		err := tx.NewSelect().
+			Column("id").
+			Table("projects").
+			Where("key = ?", key).
+			Scan(ctx, &projectID)
+		if err != nil && errors.Is(err, sql.ErrNoRows) {
+			return db.ErrNotFound
+		} else if err != nil {
+			return err
+		}
+
+		project, err = GetProjectByID(ctx, projectID)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	// GetProjectByID handles the case where the project is not found.
+	if err != nil {
+		return nil, err
+	}
+	return project, nil
+}
+
+// GenerateProjectKey generates a unique project key for a project based on its name.
+func generateProjectKey(ctx context.Context, tx bun.Tx, projectName string) (string, error) {
+	var key string
+	found := true
+	for i := 0; i < MaxRetries && found; i++ {
+		prefixLength := min(len(projectName), MaxProjectKeyPrefixLength)
+		prefix := projectName[:prefixLength]
+		suffix := random.String(MaxProjectKeyLength - prefixLength)
+		key = strings.ToUpper(prefix + suffix)
+		err := tx.NewSelect().Model(&model.Project{}).Where("key = ?", key).For("UPDATE").Scan(ctx)
+		found = err == nil
+	}
+	if found {
+		return "", fmt.Errorf("could not generate a unique project key")
+	}
+	return key, nil
+}
+
+// InsertProject inserts a new project into the database.
+func InsertProject(
+	ctx context.Context,
+	p *model.Project,
+	requestedKey *string,
+) (err error) {
+RetryLoop:
+	for i := 0; i < MaxRetries; i++ {
```

Comment:
May I please ask what was the thought behind retrying here given that we already retried in `generateProjectKey`?

Reply:
We retry in the `generateProjectKey` case to identify a project key that hasn't been used yet; however, if there are multiple request that come in we'll potentially run into a serializability error if there are concurrent inserts (i.e. two request generate the same project key and attempt to insert at the same time).